Add serialization, option creation to Blaze vector and matrix types

nilsvu opened the pull request:
## Proposed changes

Add interoperability with Blaze. This will allow working with Blaze vector and matrix types more directly. For example, I will need to use sparse linear operations a lot in optimizations of the elliptic solver, which reduce runtime of an elliptic solve by ~30%. Specifically, I need to store (and hence serialize/deserialize) incomplete LU decompositions of linear operators on each element. I currently use [Eigen3](https://eigen.tuxfamily.org) for this on my branch, but I'm considering to port the functionality to Blaze.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe commented on src/DataStructures/CompressedMatrix/OptionCreation.hpp:
```diff
@@ -0,0 +1,36 @@
+  template <typename Metavariables>
+  static blaze::CompressedMatrix<Type, SO, Tag> create(
+      const Options::Option& options) {
+    const auto data = options.parse_as<std::vector<std::vector<Type>>>();
```
`#include <vector>`

wthrowe commented on src/DataStructures/CompressedMatrix/OptionCreation.hpp:
```diff
@@ -0,0 +1,36 @@
+  static blaze::CompressedMatrix<Type, SO, Tag> create(
+      const Options::Option& options) {
+    const auto data = options.parse_as<std::vector<std::vector<Type>>>();
+    const size_t num_rows = data.size();
```
`#include <cstddef>`

wthrowe commented on src/DataStructures/CompressedMatrix/OptionCreation.hpp:
```diff
@@ -0,0 +1,36 @@
+                    "All matrix columns must have the same size.");
+      }
+      for (size_t j = 0; j < num_cols; j++) {
+        if (gsl::at(row, j) != 0.) {
```
`#include "Utilities/Gsl.hpp"`

wthrowe commented on src/DataStructures/CompressedMatrix/Serialization.hpp:
```diff
@@ -0,0 +1,51 @@
+/// Serialization of blaze::CompressedMatrix
+template <typename Type, bool SO, typename Tag>
+void pup(er& p, blaze::CompressedMatrix<Type, SO, Tag>& t) {
+  size_t rows = t.rows();
```
Go through everything in the PR and check the includes.

wthrowe commented on src/DataStructures/CompressedMatrix/OptionCreation.hpp:
```diff
@@ -0,0 +1,36 @@
+      const auto& row = gsl::at(data, i);
+      if (row.size() != num_cols) {
+        PARSE_ERROR(options.context(),
+                    "All matrix columns must have the same size.");
```
columns -> rows

wthrowe commented on src/DataStructures/DynamicMatrix/OptionCreation.hpp:
```diff
@@ -0,0 +1,35 @@
+      const auto& row = gsl::at(data, i);
+      if (row.size() != num_cols) {
+        PARSE_ERROR(options.context(),
+                    "All matrix columns must have the same size.");
```
columns -> rows

wthrowe commented on src/DataStructures/StaticMatrix/OptionCreation.hpp:
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <array>
+#include <blaze/math/DynamicMatrix.h>
```
StaticMatrix?

wthrowe commented on src/DataStructures/StaticMatrix/Serialization.hpp:
```diff
@@ -0,0 +1,31 @@
+
+#pragma once
+
+#include <blaze/math/DynamicMatrix.h>
```
StaticMatrix?

wthrowe commented on tests/Unit/DataStructures/Test_BlazeInteroperability.cpp:
```diff
@@ -0,0 +1,111 @@
+    test_serialization(blaze::CompressedVector<double>{0., 1., 0., 2.});
+    CHECK(TestHelpers::test_creation<blaze::CompressedVector<double>>(
+              "[0., 1., 0., 2.]") ==
+          blaze::CompressedVector<double>{0., 1., 0., 2.});
```
For the sparse types check that the option-created and deserialized objects are sparse.

nilsvu replied:
I believe this is already tested with the equality operators, since the `blaze::CompressedMatrix/Vector` constructor only stores non-zeros. I have so far not been able to put a stored zero into a sparse matrix at all, which gives me confidence that Blaze is doing a good job :) If you want me to add another test nevertheless, just let me know.

wthrowe replied:
I believe Blaze `==` tests for mathematical equivalence, not internal-representation equivalence.  The test still passes if I change CompressedVector to DynamicVector on this line.

More generally, you don't trust Blaze to get it right in the main code.  (There are checks for zero entries, which is what I wanted a test of.)  If we can trust here Blaze we don't need those checks.  If we can't then we shouldn't be more trusting in the test.

nilsvu replied:
I added tests in a fixup, please take a look. I don't see a lower-level way of testing the internal representation.

wthrowe replied:
That's what I had in mind.  Squash.

nilsdeppe reviewed (changes requested):
I really dislike the directory structure this is adding. I think that's more complicated than it needs to be.
1. Why would we want to create matrices in input files?
2. Why do we need to create every different type of matrix in an input file?
3. We don't create any DataVectors in input files, so why start adding support for all this now?
4. I don't understand how any of these changes related to the 30% performance improvement that using sparse matrices in the elliptic solver gains. The PR comment doesn't make it clear to me at all.

nilsvu committed:
Add serialization, option creation to Blaze vector and matrix types

nilsvu commented:
@wthrowe @nilsdeppe I removed the directory structure and created a single file `DataStructures/CompressedMatrix.hpp` etc for each type. I agree this makes it easier to work with the Blaze matrix and vector types. Re @nilsdeppe's questions:

- Some linear solver tests create matrices and vectors from input files (see e.g. `Test_GmresAlgorithm.yaml` and friends). This currently only works with the `DenseMatrix/Vector` thin wrapper classes, but I think with this PR and #3806 we can get rid of the wrappers.
- What I need for the elliptic solver is support for serialization of the Blaze types (mostly CompressedMatrix and DynamicVector, maybe HybridMatrix later once I've had the chance to benchmark it). This is used to cache incomplete LU decompositions of linear operators on each element. I currently use Eigen for this, but it would be nice and possibly even faster to do it all in Blaze. Added some context to the PR description.

nilsdeppe commented on src/DataStructures/CompressedMatrix.hpp:
```diff
@@ -0,0 +1,89 @@
+      p | num_non_zeros;
+      for (size_t j = 0; j < num_non_zeros; ++j) {
+        p | index;
+        if constexpr (SO == blaze::rowMajor) {
```
This is a surprising implementation to me. Naively I expected something like `p | t.data();`. Charm++ used to have serialization of `std::vector` be done element-by-element and it was terribly slow. If there's a way to do this across a contiguous data set with `PUP_array_bytes` then it'll be a lot faster, except if you have a sparse matrix with way more data storage than data, but presumably those cases should be avoided in general :) If there isn't a way to do a contiguous stream, could you please add a brief comment why not so that in the future if someone looks at this implementation they'll at least know it's intentional?

nilsvu replied:
Pushed a fixup that switches the serialization to use Blaze's low-level `append` mechanism. We can't use `PUParray` because the sparse types don't expose `.data()`, since their storage is a combination of values and indices. I added a comment that there may be a faster way to do this copying raw memory, but I haven't figured it out yet.

nilsdeppe commented on src/DataStructures/CompressedVector.hpp:
```diff
@@ -0,0 +1,64 @@
+    const auto data = options.parse_as<std::vector<T>>();
+    blaze::CompressedVector<T, TF, Tag> result(data.size());
+    for (size_t i = 0; i < data.size(); ++i) {
+      if (gsl::at(data, i) != 0.) {
```
could you add a comment why not just use `std::copy` (I understand why, but better to communicate to sleepy Nils in the future :) )

nilsdeppe reviewed (commented):
Few minor comments, squash immediately.

Thanks for the explanation! That all sounds great to me :)

nilsdeppe reviewed (dismissed):
LGTM, please squash! No need for my explicit re-approval to merge!
